source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+use core::any::TypeId;
+
+use mnemos_alloc::{containers::HeapFixedVec, heap::HeapGuard};
+use postcard::experimental::max_size::MaxSize;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use uuid::{uuid, Uuid};
+
+use crate::comms::{
+    bbq,
+    kchannel::{KProducer, LeakedKProducer},
+    rosc::Sender,
+};
+
+/// A partial list of known UUIDs of driver services
+pub static KNOWN_UUIDS: &[Uuid] = &[
+    //
+    // Kernel UUIDs
+    //
+
+    // SerialMux
+    uuid!("54c983fa-736f-4223-b90d-c4360a308647"),
+    // Simple Serial Port
+    uuid!("f06aac01-2773-4266-8681-583ffe756554"),
+];
+
+/// A marker trait designating a registerable driver service.
+///
+/// Typically used with [Registry::set] or [Registry::set_konly].
+/// Can typically be retrieved by [Registry::get] or [Registry::get_userspace]
+/// After the service has been registered.
+pub trait RegisteredDriver {
+    /// This is the type of the request sent TO the driver service
+    type Request: 'static;
+
+    /// This is the type of the response sent FROM the driver service
+    type Response: 'static;
+
+    /// This is the UUID of the driver service
+    const UUID: Uuid;
+
+    /// Get the type_id used to make sure that driver instances are correctly typed.
+    /// Corresponds to the same type ID as `(Self::Request, Self::Response)`
+    fn type_id() -> TypeId {
+        TypeId::of::<(Self::Request, Self::Response)>()
+    }
+}
+
+/// The driver registry used by the kernel.
+pub struct Registry {
+    items: HeapFixedVec<RegistryItem>,
+}
+
+// TODO: This probably goes into the ABI crate, here is fine for now
+#[derive(Serialize, Deserialize)]
+pub struct UserRequest<'a> {
+    // TODO: Maybe not the UUID, maybe pre-discover a shorter UID?
+    uid: Uuid,
+    nonce: u32,
+    #[serde(borrow)]
+    req_bytes: &'a [u8],
+}
+
+// TODO: This probably goes into the ABI crate, here is fine for now
+#[derive(Serialize, Deserialize)]
+pub struct UserResponse<U> {
+    // TODO: Maybe not the UUID, maybe pre-discover a shorter UID?
+    uuid: Uuid,
+    nonce: u32,
+    reply: Result<U, ()>,
+}
+
+/// A wrapper for a message TO and FROM a driver service.
+/// Used to be able to add additional message metadata without
+/// changing the fundamental message type.
```

Comment:
does that mean that the intention is to add more fields to the `Envelope` type? if so, it should probably be
```suggestion
/// Used to be able to add additional message metadata without
/// changing the fundamental message type.
#[non_exhaustive]
```